src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -7438,6 +7460,224 @@ private BoundExpression BindIndexedPropertyAccess(SyntaxNode syntax, BoundExpres
+                lookupResult.Clear();
+                useSiteDiagnostics?.Clear();
+                LookupMembersInType(
+                    lookupResult,
+                    receiverType,
+                    WellKnownMemberNames.CountPropertyName,
+                    arity: 0,
+                    basesBeingResolved: null,
+                    LookupOptions.Default,
+                    originalBinder: this,
+                    diagnose: false,
+                    ref useSiteDiagnostics);
+
+                if (!lookupSatisfiesLengthPattern(out lengthOrCountProperty))
+                {
+                    cleanup();
+                    patternIndexerAccess = null;
+                    return false;
+                }
+            }
+
+            Debug.Assert(lengthOrCountProperty is { });
+            lookupResult.Clear();
+            useSiteDiagnostics?.Clear();
+
+            if (argIsIndex)
+            {
+                // Look for `T this[int i]` indexer
+
+                LookupMembersInType(
+                    lookupResult,
+                    receiverType,
+                    WellKnownMemberNames.Indexer,
+                    arity: 0,
+                    basesBeingResolved: null,
+                    LookupOptions.Default,
+                    originalBinder: this,
+                    diagnose: false,
+                    ref useSiteDiagnostics);
+
+                if (lookupResult.IsMultiViable)
+                {
+                    foreach (var candidate in lookupResult.Symbols)
+                    {
+                        if (!candidate.IsStatic &&
+                            IsAccessible(candidate, ref useSiteDiagnostics) &&
+                            candidate is PropertySymbol property &&
+                            property.GetOwnOrInheritedGetMethod() is MethodSymbol getMethod &&
+                            getMethod.OriginalDefinition is var original &&
+                            original.ParameterCount == 1 &&
+                            isIntNotByRef(original.Parameters[0]))
+                        {
+                            patternIndexerAccess = new BoundIndexOrRangePatternIndexerAccess(
+                                syntax,
+                                receiverOpt,
+                                lengthOrCountProperty,
+                                getMethod,
+                                arguments[0],
+                                getMethod.ReturnType);
+                            cleanup();
+                            return true;
+                        }
+                    }
+                }
+            }
+            else if (receiverType.SpecialType == SpecialType.System_String)
+            {
+                Debug.Assert(argIsRange);
+                // Look for Substring
+                var substring = (MethodSymbol)Compilation.GetWellKnownTypeMember(WellKnownMember.System_String__Substring);
+                if (substring is { })
```

Comment:
Super nit: do we need to check for accessibility here?

Reply:
WellKnownMembers are always public. That's maybe slightly more restrictive than necessary, but doesn't seem like a problem.